javascript/lifecycle.js

```diff
@@ -125,56 +186,51 @@ document.addEventListener(
+  reflexId
+) => {
+  if (!controllerElement) {
     if (Debug.enabled && !reflexes[reflexId].warned) {
       console.warn(
-        `StimulusReflex was not able execute the "${stage}" or later life-cycle methods on the element which triggered the Reflex. The element is no longer present in the DOM. Could you move the Reflex action to an element higher in your DOM?`
+        `StimulusReflex was not able execute callbacks or emit events for "${stage}" or later life-cycle stages for this Reflex. The StimulusReflex Controller Element is no longer present in the DOM. Could you move the StimulusReflex Controller to an element higher in your DOM?`
       )
       reflexes[reflexId].warned = true
     }
     return
   }
 
-  const reflexData = element.reflexData || {}
-  const reflexController = element.reflexController || {}
-  const reflexError = element.reflexError || {}
-  const oldElement = element
-
-  if (!document.body.contains(element)) {
-    const attrs = extractElementAttributes(element)
-    element = findElement(attrs)
-
-    if (Debug.enabled)
-      console.warn(
-        `StimulusReflex detected that the element which triggered the Reflex has been replaced by a morph operation. If you rely on all life-cycle methods to be executed, move the Reflex action to an element higher in your DOM.`
-      )
-  }
-
-  if (!element) {
-    if (Debug.enabled && !reflexes[reflexId].warned) {
-      console.warn(
-        `StimulusReflex was not able execute the "${stage}" or later life-cycle methods on the element which triggered the Reflex. The following element is no longer present in the DOM: `,
-        oldElement
-      )
-      reflexes[reflexId].warned = true
-    }
-    return
+  if (
+    !controllerElement.reflexController &&
+    Debug.enabled &&
+    !reflexes[reflexId].warned
+  ) {
+    console.warn(
+      `StimulusReflex detected that the StimulusReflex Controller responsible for this Reflex has been replaced with a new instance. Callbacks and events for "${stage}" or later life-cycle stages cannot be executed.`
+    )
+    reflexes[reflexId].warned = true
   }
 
-  element.reflexData = reflexData
-  element.reflexController = reflexController
-  element.reflexError = reflexError
-
-  const { target } = element.reflexData[reflexId] || {}
-  const { controller } = element.reflexController[reflexId] || {}
+  const { target } = controllerElement.reflexData[reflexId] || {}
+  const controller = controllerElement.reflexController[reflexId] || {}
   const event = `stimulus-reflex:${stage}`
-  const detail = { reflex: target, controller, reflexId }
+  const detail = {
+    reflex: target,
+    controller,
+    reflexId,
+    element: reflexElement
+  }
 
-  element.dispatchEvent(
+  controllerElement.dispatchEvent(
     new CustomEvent(event, { bubbles: true, cancelable: false, detail })
   )
-  if (window.jQuery) window.jQuery(element).trigger(event, detail)
+  if (window.jQuery) window.jQuery(controllerElement).trigger(event, detail)
```

Comment:
Shouldn't we dispatch lifecycle events on the `reflexElement` given that the event will bubble up to the controller if someone wants to handle it there?

Reply:
Yes, that's an exceptionally smart call.